### PR TITLE
Standardized in-app authentication

### LINKED ISSUE #15570: Implement standardized in app authentication process
Design team are currently working on a standardised authentication process for in app use. 

Designs: 
https://www.figma.com/file/V6nlpAWIf2e1XU8RJ9yQPe/Syncing-for-Mobile?node-id=1104-155973&t=CjF08I1JH0ASv6JA-0

https://www.figma.com/file/zG04IdiVCb4ZNC5n3nH9yg/Authentication-for-Mobile?type=design&node-id=0%3A1&t=nnuEZbGYPpwGAGcW-1

The solution should also work for these pages- https://www.figma.com/file/HKncH4wnDwZDAhc4AryK8Y/Wallet-for-Mobile?type=design&node-id=1741-374592&mode=design&t=FGZZnmf1pJQzj0d7-0
 i.e it is a generic approach and has the required callbacks for success/failure states etc 

Double check with design team these are ready to go before starting 👍 

**Keycard variant is out of scope!**

### PR DESCRIPTION
**Pull request by hysekbutt**
fixes #15570 

### Summary

- Auth process including face-id/touch-id.
- If biometrics aren't available the fallback method of authentication, i.e. password authentication will be used.
- Done the changes in UI to match the Figma designs.

#### Platforms

- Android
- iOS

### Steps to test

- Open Status
- Navigate to `Syncing > Slide to reveal > Enter Password`

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -6719,7 +6719,7 @@ js-yaml@^3.13.1:
     argparse "^1.0.7"
     esprima "^4.0.0"
 
-jsc-android@^250230.2.1:
+jsc-android@250230.2.1, jsc-android@^250230.2.1:
```
Comment (J-Son89) [inline yarn.lock:6722]: probably can be reverted 👍

File: ios/Podfile.lock [hunk 2]
```diff
@@ -772,7 +772,7 @@ SPEC CHECKSUMS:
   RNLanguages: 962e562af0d34ab1958d89bcfdb64fafc37c513e
   RNPermissions: ad71dd4f767ec254f2cd57592fbee02afee75467
   RNReactNativeHapticFeedback: 2566b468cc8d0e7bb2f84b23adc0f4614594d071
```
Comment (J-Son89) [inline ios/Podfile.lock:774]: probably can be reverted 👍

File: src/quo2/components/buttons/button/style.cljs [hunk 3]
```diff
@@ -103,4 +103,4 @@
            {:border-color border-color
             :border-width 1})
          (when disabled
-           {:opacity 0.3})))
+           {:opacity 0.7})))
```
Comment (J-Son89) [inline src/quo2/components/buttons/button/style.cljs]: curious of why did you do this? these styles were set correctly - button is used on pretty much every page so it's best to reach out to be sure if there's some changes like this needed

File: src/quo2/components/buttons/slide_button/view.cljs [hunk 4]
```diff
@@ -55,14 +55,14 @@
           [reanimated/view {:style (style/track-cover interpolate-track)}
            [rn/view {:style (style/track-cover-text-container @track-width)}
             [icon/icon track-icon
-             {:color (utils/slider-color :main customization-color)
```
Comment (J-Son89) [inline src/quo2/components/buttons/slide_button/view.cljs]: `customization-color` was most likely correct - are you sure of this here? 
Did you understand how `customition-color` should be working? if there's any doubts feel free to reach out 👍

File: src/status_im2/contexts/syncing/events.cljs [hunk 5]
```diff
@@ -72,14 +72,31 @@
                          :event    :syncing/input-connection-string-for-bootstrapping}))))]
     (native-module/prepare-dir-and-update-config "" default-node-config-string callback)))
 
+;; TODO :: Password verification is currently based on length but we have to do it realtime
```
Comment (J-Son89) [inline src/status_im2/contexts/syncing/events.cljs]: let's remove dead code. you could add it as a comment in the pr review if it's necessary to communicate this out 👍

File: src/status_im2/contexts/syncing/setup_syncing/view.cljs [hunk 6]
```diff
@@ -38,30 +38,30 @@
...
-                                                   (- (* 1000
-                                                         (js/Math.ceil (/ (datetime/timestamp) 1000)))
-                                                      @timestamp))))
-                          (reset! delay nil)))
-        cleanup-clock (fn []
-                        (reset! code nil)
-                        (reset! timestamp nil)
-                        (reset! valid-for-ms code-valid-for-ms))]
+  (let [valid-for-ms                           (reagent/atom code-valid-for-ms)
+        {:keys [customization-color]}          (rf/sub [:profile/multiaccount])
```
Comment (J-Son89) [inline src/status_im2/contexts/syncing/setup_syncing/view.cljs]: this sub looks wrong as syncing is both in app and a login method.
I recently created a `sub` which should work for these pages
you can use 
`customization-color (rf/sub [:profile/customization-color])`

File: src/status_im2/contexts/syncing/setup_syncing/style.cljs [hunk 7]
```diff
@@ -1,5 +1,6 @@
 (ns status-im2.contexts.syncing.setup-syncing.style
-  (:require [quo2.foundations.colors :as colors]))
+  (:require [quo2.foundations.colors :as colors]
+            [status-im2.contexts.syncing.standard-authentication.view :as standard-auth]))
```
Comment (J-Son89) [inline src/status_im2/contexts/syncing/setup_syncing/style.cljs]: looks like this `require` - "[status-im2.contexts.syncing.standard-authentication.view :as standard-auth]" is not needed

File: src/status_im2/contexts/syncing/events.cljs [hunk 8]
```diff
@@ -72,14 +72,31 @@
...
+
 (rf/defn preparations-for-connection-string
   {:events [:syncing/get-connection-string-for-bootstrapping-another-device]}
   [{:keys [db]} entered-password set-code]
   (let [valid-password? (>= (count entered-password) constants/min-password-length)
         show-sheet      (fn [connection-string]
                           (set-code connection-string)
                           (rf/dispatch [:syncing/update-role constants/local-pairing-role-sender])
-                          (rf/dispatch [:bottom-sheet/hide]))]
+                          (rf/dispatch [:hide-bottom-sheet]))]
```
Comment (J-Son89) [inline src/status_im2/contexts/syncing/events.cljs:82]: @smohamedjavid this is the right call for bottom sheet methods? 🤔

File: src/quo2/components/tags/context_tag/style.cljs [hunk 9]
```diff
@@ -21,15 +21,16 @@
...
+   :font-size      13})
 
 (def audio-tag-container
   {:padding-left     2
    :padding-vertical 2})
 
 (def audio-tag-icon-container
-  {:width            20
-   :height           20
+  {:width            24
```
Comment (J-Son89) [inline src/quo2/components/tags/context_tag/style.cljs]: hmm, why did you need to update this component? 🤔

File: src/status_im2/contexts/syncing/setup_syncing/style.cljs [hunk 10]
```diff
@@ -46,3 +60,13 @@
    :left              0
    :right             0
    :margin-horizontal 60})
+
+(def title-container-22
```
Comment (J-Son89) [inline src/status_im2/contexts/syncing/setup_syncing/style.cljs]: why `-22` ? seem like an implementation detail that's not relevant to the name

File: src/status_im2/contexts/syncing/setup_syncing/style.cljs [hunk 11]
```diff
@@ -46,3 +60,13 @@
...
+  {:flex-direction      :row
+   :align-items         :center
+   :border-radius       20
+   :padding-horizontal  20
+   :padding-vertical    12
+   :margin-top          12
+   :justify-content     :space-between
+   :background-color    "rgba(42, 74, 245, 0.1)"})
```
Comment (J-Son89) [inline src/status_im2/contexts/syncing/setup_syncing/style.cljs]: not sure were you are getting this color but in 99% of use cases the colors should be in `quo2.foundations.colors`

File: src/status_im2/contexts/syncing/setup_syncing/view.cljs [hunk 12]
```diff
@@ -130,6 +114,20 @@
                :container-style {:margin-top 12}
                :icon-left       :i/copy}
               (i18n/label :t/copy-qr)]])]]
+        (when-not (sync-utils/valid-connection-string? @code)
+          ;; Implemented standard in-app authentication
```
Comment (J-Son89) [inline src/status_im2/contexts/syncing/setup_syncing/view.cljs]: imo we can remove this comment 👍

File: src/status_im2/contexts/syncing/setup_syncing/view.cljs [hunk 13]
```diff
@@ -130,6 +114,20 @@
...
+          [rn/view {:style style/standard-auth}
+           [standard-auth/view
+            {:track-text              (i18n/label :t/slide-to-reveal-code)
+             :thumb-color             profile-color
```
Comment (J-Son89) [inline src/status_im2/contexts/syncing/setup_syncing/view.cljs]: we refer to the "users" color as `customization-color` across the codebase - this should be the same here 👍

File: src/status_im2/contexts/syncing/standard_authentication/style.cljs [hunk 14]
```diff
@@ -0,0 +1,14 @@
+(ns status-im2.contexts.syncing.standard-authentication.style)
```
Comment (J-Son89) [inline src/status_im2/contexts/syncing/standard_authentication/style.cljs:1]: I notice that you are using this `style` namespace in 2 components. We generally try to use 1 style file per view file and if styles need to be reused it is better to do with a component rather and if not possible it is generally better to just duplicate so we are not putting unnecessary coupling between two indeendeny views and their styles

File: src/status_im2/contexts/syncing/standard_authentication/view.cljs [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+  (:require
+   [quo2.core :as quo]
+   [utils.re-frame :as rf]
+   [react-native.touch-id :as biometric]
+   [utils.i18n :as i18n]
+   [status-im2.contexts.syncing.standard-authentication.enter-password :as enter-password]
+   [status-im2.contexts.syncing.standard-authentication.style :as style]
+   [react-native.core :as rn]))
+
+(defn authorize [set-code use-biometric-auth auth-success auth-fail fallback-button-label]
```
Comment (J-Son89) [inline src/status_im2/contexts/syncing/standard_authentication/view.cljs]: imo better to use a map here
```
(defn authorize [{:keys [set-code use-biometric-auth auth-success auth-fail fallback-button-label]}]
```

File: src/status_im2/contexts/syncing/standard_authentication/view.cljs [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+         :on-success    (fn [response]
+                          (auth-success)
+                          (js/console.log "response" response))
+         :on-fail       (fn [error]
+                          (js/console.log "Authentication Failed. Error:" error)
+                          (auth-fail)
+                          (rf/dispatch [:show-bottom-sheet
+                                        {:content (fn [] [enter-password/view set-code])}]))
+         :options       {}})
+           ;; If biometric authentication is not supported, fallback to password-based authentication
```
Comment (J-Son89) [inline src/status_im2/contexts/syncing/standard_authentication/view.cljs]: imo: I prefer to avoid comments like this and just use named functions which describe the logic 
```
(let [biometric-supported? (and use-biometric-auth biometric-type)]
 (if (not biometric-supported?)
   (biometric-auth {...})
   (password-auth {...})
 )))
 
```

File: src/status_im2/contexts/syncing/standard_authentication/view.cljs [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+           enter-pass-callback
+           use-biometric-auth
+           on-auth-success
+           on-auth-fail]}]
+  [rn/view {:style style/main-container}
+   [quo/slide-button
+    {:customization-color   customization-color
+     :on-complete           #(authorize enter-pass-callback use-biometric-auth on-auth-success on-auth-fail fallback-button-label)
+     :thumb-color           thumb-color
+     :track-icon            :face-id
```
Comment (J-Son89) [inline src/status_im2/contexts/syncing/standard_authentication/view.cljs]: icons should be qualified with `:i/...`. i.e `:track-icon :i/face-id`

File: src/status_im2/contexts/syncing/standard_authentication/view.cljs [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+     :track-text            track-text
+     :track-text-color      track-text-color}]])
```
Comment (J-Son89) [inline src/status_im2/contexts/syncing/standard_authentication/view.cljs]: as a general rule our design system components are encapsulating colors and styles, so if we're setting colors like this then it suggest something is wrong with the underlyinh component - I think it's related to the use of `customization-color` mentioned elsewhere in this pr.

File: src/status_im2/contexts/syncing/events.cljs (hunk 8 above)
Comment (smohamedjavid) [reply src/status_im2/contexts/syncing/events.cljs:82]: `:bottom-sheet/hide` is an old bottom sheet code & event - this needs to be cleaned up with this implementation of authentication.

`:hide-bottom-sheet` - new bottom sheet event and it's correct.

File: src/status_im2/contexts/syncing/standard_authentication/style.cljs [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+
+(def main-container
+  {:width  "100%"})
+
+(def enter-password-container
+  {:margin-horizontal        20
+   :border-top-left-radius   12
+   :border-top-right-radius  12})
```
Comment (briansztamfater) [inline src/status_im2/contexts/syncing/standard_authentication/style.cljs]: There's extra spacing in map columns, probably `make lint-fix` should fix these kind of formatting issues

File: src/quo2/components/buttons/slide_button/view.cljs (hunk 4 above)
Comment (hysekbutt) [reply src/quo2/components/buttons/slide_button/view.cljs]: Yes, I'm pretty sure about it.

File: src/status_im2/contexts/syncing/standard_authentication/style.cljs (hunk 14 above)
Comment (hysekbutt) [reply src/status_im2/contexts/syncing/standard_authentication/style.cljs:1]: Now, It's being used in 1 component.

File: src/quo2/components/tags/context_tag/style.cljs (hunk 9 above)
Comment (hysekbutt) [reply src/quo2/components/tags/context_tag/style.cljs]: Was trying to give the styling as per figma designs

File: src/status_im2/contexts/syncing/standard_authentication/enter_password.cljs [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+
+       [rn/view
+        {:margin-top 45}
+        [quo/button
+         {:size                40
+          :type                :primary
+          :customization-color (or customization-color :primary)
+          :accessibility-label :login-button
+          :before              :i/reveal
+          :disabled            (or (not sign-in-enabled?) processing)
```
Comment (J-Son89) [inline src/status_im2/contexts/syncing/standard_authentication/enter_password.cljs]: There was some recent changes to button that got merged.

':before' -> ':icon-left'
':disabled' -> ':disabled?'
':style' -> ':container-style'

Most likely you can revert the changes you made to the button component as that would break all other uses in the app

File: src/status_im2/contexts/syncing/standard_authentication/view.cljs [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+           on-auth-fail]}]
+  [rn/view {:style {:width "100%"}}
+   [quo/slide-button
+    {:customization-color customization-color
+     :on-complete         #(authorize {:set-code              enter-pass-callback
+                                       :use-biometric-auth    use-biometric-auth
+                                       :auth-success          on-auth-success
+                                       :auth-fail             on-auth-fail
+                                       :fallback-button-label fallback-button-label})
+     :thumb-color         thumb-color
```
Comment (J-Son89) [inline src/status_im2/contexts/syncing/standard_authentication/view.cljs]: this seems wrong as we like to encapsulate the colors of the design system components within the component - with the exception of setting cusotomization-color with the `key` from the users choice, e.g `:blue` `:army` etc.

It is better to make those adjustments within the component directly so they do not need to be set.

In this case I checked the design system and there is a discrepancy in these colors between the designs and the syncing flows - so I reached out to design to see which version is correct, as they need to align this 👍 

https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System-for-Mobile?type=design&node-id=19085-205204&mode=design&t=htgflfZwaUfqzAMW-4

File: package.json [hunk 22]
```diff
@@ -34,6 +34,7 @@
     "eth-phishing-detect": "^1.2.0",
     "functional-red-black-tree": "^1.0.1",
     "i18n-js": "^3.3.0",
+    "jsc-android": "250230.2.1",
```
Comment (siddarthkay) [inline package.json]: Hmm, I'm curious @hysekbutt , why was this included? Can you elaborate a bit please?

File: src/status_im2/contexts/syncing/standard_authentication/enter_password.cljs [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+       [quo/text
+        {:accessibility-label :sync-code-generated
+         :weight              :bold
+         :size                :heading-1
+         :style               {:color         (colors/theme-colors colors/neutral-100 colors/white)
+                               :margin-bottom 4}}
+        (i18n/label :t/enter-password)]
+
+       [rn/view
+        {:flex-direction :row :margin-bottom 20 :margin-top 8}
```
Comment (siddarthkay) [inline src/status_im2/contexts/syncing/standard_authentication/enter_password.cljs]: move these to dedicated style file please.
ref : https://github.com/status-im/status-mobile/blob/develop/doc/new-guidelines.md#component-styles

File: src/status_im2/contexts/syncing/standard_authentication/enter_password.cljs [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+        [quo/context-tag {:blur? false} profile-picture display-name]]
+
+       [rn/view {:flex-direction :row :align-items :flex-end}
+        [rn/view {:flex 6}
```
Comment (siddarthkay) [inline src/status_im2/contexts/syncing/standard_authentication/enter_password.cljs]: move these to dedicated style file please.
ref : https://github.com/status-im/status-mobile/blob/develop/doc/new-guidelines.md#component-styles

File: src/status_im2/contexts/syncing/standard_authentication/enter_password.cljs [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+           {:type :error
+            :size :default
+            :icon :i/info}
+           error]
+          [rn/touchable-opacity
+           {:hit-slop       {:top 6 :bottom 20 :left 0 :right 0}
+            :disabled       false
+            :active-opacity 1
+            :on-press       #(reset! entered-password %)}
+           [rn/text
```
Comment (siddarthkay) [inline src/status_im2/contexts/syncing/standard_authentication/enter_password.cljs]: maybe this should be `[quo/text` instead of `[rn/text`

File: src/status_im2/contexts/syncing/standard_authentication/enter_password.cljs [hunk 26]
```diff
@@ -0,0 +1,97 @@
...
+       [rn/view
+        {:margin-top 45}
+        [quo/button
+         {:size                40
+          :type                :primary
+          :customization-color (or customization-color :primary)
+          :accessibility-label :login-button
+          :before              :i/reveal
+          :disabled            (or (not sign-in-enabled?) processing)
+          :style               {:margin-bottom 12 :background-color "#5A33CA"}
```
Comment (siddarthkay) [inline src/status_im2/contexts/syncing/standard_authentication/enter_password.cljs]: move these to dedicated style file please.
ref : https://github.com/status-im/status-mobile/blob/develop/doc/new-guidelines.md#component-styles

File: src/status_im2/contexts/syncing/standard_authentication/view.cljs (hunk 21 above)
Comment (siddarthkay) [reply src/status_im2/contexts/syncing/standard_authentication/view.cljs]: I agree @J-Son89

Comment (J-Son89) [reply src/status_im2/contexts/syncing/standard_authentication/view.cljs]: ah, so as it turns out this variant using `blur?` being active is not created in the component yet. 
I would prefer that we do not hack the component like this and instead it can just be a note to be taken into consideration when the designers are looking at the component.  👍 
I will create the follow up issue to fix the slider button with the `blur` variant.
Screenshot 2023-08-04 at 13 01 12

File: src/status_im2/contexts/syncing/standard_authentication/enter_password.cljs (hunk 24 above)
Comment (briansztamfater) [reply src/status_im2/contexts/syncing/standard_authentication/enter_password.cljs]: By the way, why `:flex 6`?

File: src/quo2/components/buttons/button/style.cljs [hunk 27]
```diff
@@ -53,7 +53,7 @@
                       12))})
 
 (defn style-container
-  [{:keys [size disabled? border-radius background-color border-color icon-only? icon-top
+  [{:keys [size disabled border-radius background-color border-color icon-only? icon-top
```
Comment (briansztamfater) [inline src/quo2/components/buttons/button/style.cljs]: Why is question mark removed from `disabled?` prop?

File: src/status_im2/contexts/syncing/standard_authentication/enter_password.cljs (hunk 20 above)
Comment (hysekbutt) [reply src/status_im2/contexts/syncing/standard_authentication/enter_password.cljs]: updated these props where I'm using this component.

File: src/status_im2/contexts/syncing/standard_authentication/view.cljs (hunk 21 above)
Comment (hysekbutt) [reply src/status_im2/contexts/syncing/standard_authentication/view.cljs]: Yes, that's why I was passing down the colors. For now I have reverted the changes that I did in the component.

File: package.json (hunk 22 above)
Comment (hysekbutt) [reply package.json]: reverted

File: src/quo2/components/buttons/button/style.cljs (hunk 27 above)
Comment (hysekbutt) [reply src/quo2/components/buttons/button/style.cljs]: reverted.

File: yarn.lock [hunk 28]
```diff
@@ -6724,7 +6724,7 @@ js-yaml@^3.13.1:
     argparse "^1.0.7"
     esprima "^4.0.0"
 
-jsc-android@^250230.2.1:
```
Comment (siddarthkay) [inline yarn.lock]: I see you removed an entry from `package.json` it is also better to revert this file as well @hysekbutt, since nix relies on `yarn.lock` to generate dependencies :)

File: ios/Podfile.lock [hunk 29]
```diff
@@ -425,8 +425,8 @@ PODS:
     - React-Core
   - RNStaticSafeAreaInsets (2.2.0):
     - React-Core
-  - RNSVG (9.13.6):
-    - React
+  - RNSVG (13.10.0):
+    - React-Core
```
Comment (siddarthkay) [inline ios/Podfile.lock]: Hmm I am seeing this in my branch as well, it seems someone updated the SVG library and forgot to update the `Podfile.lock` , we can do that in a separate PR, Please revert all changes to `Podfile.lock` in this case in your PR

File: yarn.lock (hunk 28 above)
Comment (awais0198) [reply yarn.lock:6727]: Removed from `yarn.lock`

Comment (hysekbutt) [reply yarn.lock]: I already removed the changes of `yarn.lock` file

**Comment by siddarthkay** [comment]
@hysekbutt : Thanks for your contribution, We have imported this PR and some test will run on CI now.

**Comment by J-Son89** [comment]
closing as a duplicate now
